Fix misread of source if path is already valid

simonrainerson opened the pull request:
# Motivation
When receiving a stream of NARs through the ssh-ng protocol, an already existing path would cause the NAR archive to not be read in the stream, resulting in trying to parse the NAR as a ValidPathInfo. This results in the error message:

    error: not an absolute path: 'nix-archive-1'

# Context
Usually this problem is avoided by running QueryValidPaths before AddMultipleToStore, but can arise when two parallel nix processes gets the same response from QueryValidPaths. This makes the problem more prominent when running builds in parallel.
Fixes #6253

<!-- Non-trivial change: Briefly outline the implementation strategy. -->

<!-- Invasive change: Discuss alternative designs or approaches you considered. -->

<!-- Large change: Provide instructions to reviewers how to read the diff. -->

# Checklist for maintainers

<!-- Contributors: please leave this as is -->

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or bug fix: updated release notes

edolstra commented:
Can you give more information on how to reproduce this? In the (current) daemon protocol, NARs are supposed to be wrapped in a `FramedSource` / `FramedSink`, which is intended specifically to allow incomplete NAR reads/writes.

simonrainerson commented:
> Can you give more information on how to reproduce this? In the (current) daemon protocol, NARs are supposed to be wrapped in a `FramedSource` / `FramedSink`, which is intended specifically to allow incomplete NAR reads/writes.

We could reproduce it consistently by running two `nix copy` with ssh-ng at the same time against a nix store in a local VM that is missing some derivation.

One succeeds and the other one gets the `error: not an absolute path: 'nix-archive-1'` error.

The full command was `NIX_SSHOPTS="-p 2221 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no" nix copy --to ssh-ng://root@localhost /nix/store/some.drv --derivation -L -vv` with the .drv file we wanted to copy.

abbec commented:
> Can you give more information on how to reproduce this? In the (current) daemon protocol, NARs are supposed to be wrapped in a `FramedSource` / `FramedSink`, which is intended specifically to allow incomplete NAR reads/writes.

It is not really an incomplete NAR parsing as much as a completely absent one. Would a `FramedSource` be able to hand let that? I mean the next read from the source would be to parse the path info since the path already existing is not really an error, right?

thufschmitt commented:
I'm actually not surprised at all (in hindsight, I have to admit I totally overlooked this): This bug had already been reported in the general case (https://github.com/NixOS/nix/issues/6730) and fixed (although I apparently forgot to close the issue) while I rewrote the implementation of `nix copy` for everything but `ssh-ng` stores (https://github.com/NixOS/nix/pull/6612). But it's logical then that `ssh-ng` stores still suffer from that bug

thufschmitt commented on src/libstore/local-store.cc:
```diff
@@ -697,6 +697,17 @@ void canonicalisePathMetaData(const Path & path,
 }
 
 
+void discardNar(Source& source, const ValidPathInfo& info) {
+    int count = info.narSize / 4096;
+    char buffer[4096];
+    for (int i=0; i < count; ++i) {
+        source(buffer, 4096);
+    }
+    int extra = info.narSize % 4096;
+    source(buffer, extra);
+}
```
Can you extract the `discard`part of it in `serialise.{cc,hh}`? (maybe even add this as a method of `Source` since it seems to be a generally useful thing to have) and just have this be a thin wrapper like `source.discard(info.narSize)`?

thufschmitt replied:
Alternatively we could maybe properly parse the Nar before discarding it. That would be marginally slower but also be more resilient in case of a corrupted stream. @edolstra , any opinion?

simonrainerson replied:
Sounds very sensible to me.

We could not come up with a "easy self-contained reproducer", it would have been nice indeed. If there's an easy way of creating a dummy store for LocalStore then we can create some data to run addToStore with and mock isValidPath so we end up in the discard code.

edolstra replied:
Yes, it's better to use
```
                    ParseSink sink; /* null sink; just parse the NAR */
                    parseDump(sink, source);
```
This also has the advantage that it doesn't require `info.narSize` to be correct. (IIRC there are some legacy cases where it can be zero.)

thufschmitt reviewed (changes requested):
I think that's very sensible :+1: 

Apart from the small code comment, do you have an easy self-contained reproducer for this? It would be nice to add as a test case

edolstra commented:
I think the real bug here is that `wopAddMultipleToStore` uses a single `FramedSource` for all paths. Instead it should use a separate `FramedSource` for each path. But that does require a protocol change.

edolstra commented on src/libstore/local-store.cc:
```diff
@@ -1343,9 +1354,15 @@ void LocalStore::addToStore(const ValidPathInfo & info, Source & source,
         outputLock.setDeletion(true);
+    } else {
+        /* Discard the NAR from the source if it has been added already. */
+        discardNar(source, info);
```
Instead of having two calls to `discardNar()`, it might be nicer to do something like:
```
bool narRead = false;
Finally cleanup = [&]() {
    if (!narRead) {
        ParseSink sink; /* null sink; just parse the NAR */
        parseDump(sink, source);
    }
};
```
and then set `narRead = true` before the call to `restorePath()`. That way it will keep working if we ever add other return paths out of this function.

abbec replied:
Yeah, that does sound much nicer

edolstra commented on doc/manual/src/release-notes/rl-next.md:
```diff
@@ -20,3 +20,10 @@
   to the embedded store and not to the host's Nix store.
 
   This requires the `discard-references` experimental feature.
+
+* Fix misread of source if path is already valid.
+  When receiving a stream of NARs through the ssh-ng protocol, an already
+  existing path would cause the NAR archive to not be read in the stream,
+  resulting in trying to parse the NAR as a ValidPathInfo. This results in
+  the error message:
+      error: not an absolute path: 'nix-archive-1'
```
```suggestion
```

since we don't list bug fixes in the release notes.

fricklerhandwerk replied:
@edolstra: Oh, but why? Except it's more work. I think long-term, large-scale users with delicate upgrade requirements may benefit from this.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/tweag-nix-dev-update-44/25546/1

thufschmitt commented:
@simonrainerson Do you think you can turn https://github.com/NixOS/nix/pull/7720#issuecomment-1412260120 into a proper test? The PR is good for me otherwise, but I'd really like this to be a bit more properly tested if that can be done

simonrainerson commented:
> @simonrainerson Do you think you can turn [#7720 (comment)](https://github.com/NixOS/nix/pull/7720#issuecomment-1412260120) into a proper test? The PR is good for me otherwise, but I'd really like this to be a bit more properly tested if that can be done

I'm not very familiar with the code so If you have any pointer to how to set up a LocalStore in a test that would be helpful. Then I think it would pretty straight forward to test for.

thufschmitt commented:
@simonrainerson I've rebased your branch and added a test on https://github.com/tweag/nix/fix-add-to-store-existing. Do you want to incorporate that into this PR? Or I can open a new one if you don't want to bother with it

simonrainerson commented:
> @simonrainerson I've rebased your branch and added a test on https://github.com/tweag/nix/fix-add-to-store-existing. Do you want to incorporate that into this PR? Or I can open a new one if you don't want to bother with it

I get a 404 from that link, feel free to open a pr if you want. I'm back at my computer at the end of the week (or start of next) and can update this PR then. Thanks either way!

thufschmitt commented:
My bad, the correct link is https://github.com/tweag/nix/tree/fix-add-to-store-existing.

I've opened https://github.com/NixOS/nix/pull/8805 as a rebased-version of this one. Should auto-merge once the CI is done running. Thanks!
